Arnatious/nodl (#3)
ubuntu-robotics/design: articles/ros2_nodl.md

These assertions results would then be summarized in a logging file for later debugging.

Another example of the usefulness of having a static interface is the ability to create graphical tools for putting a ROS system together.

> 

Reviewer: Another new paragraph to remove.

Author: 663747e

### Exporting a NoDL to the Ament Index

Per the design philosophy of the [ament index][ament_index], files will be installed in two locations.
In the case of a package named `Foo`, the NoDL file `foo.nodl.xml` should be placed in the package share directory, `share/foo/foo.nodl.xml`.
A corresponding marker file, `share/ament_index/nodl_desc/foo`, should be created.
It is either empty or contains the relative path to the `foo.nodl.xml` file.

#### CMake Macro

For packages using ament_cmake, the package `ament_nodl` provides the macro `nodl_export_node_description_file` which performs the export described in [Exporting a NoDL to the Ament Index](#exporting-a-nodl-to-the-ament-index).

For a package `Foo`, containing `foo.nodl.xml`, the following lines are added to `CMakeLists.txt`:

```cmake
find_package(ament_nodl REQUIRED)
nodl_export_node_description_file(foo.nodl.xml)
```

#### setup.py

In the case of setup.py, placement of the NoDL file and marker in the index must be done manually alongside the placement of the package's marker in the ament index.
One can re-use the same empty marker file placed in the package index.

> An example `data_files` argument to `setuptools.setup()` follows:

Reviewer 2: ```suggestion
An example `data_files` argument to `setuptools.setup()` in `setup.py` follows:
```

```python
    data_files=[
        ('share/ament_index/resource_index/packages',
            ['resource/' + package_name]),
        ('share/ament_index/resource_index/nodl_desc',
            ['resource/' + package_name]),
        ('share/' + package_name,
            ['package.xml', package_name + '.nodl.xml']),
    ],
```

### NoDL Schema

> An `.xsd` xml schema is provided alongside the NoDL implementation.

Reviewer 2: Might be worth mentioning what is that file used for.

The package interface is defined in a separate [XML][xml_wiki] file with suffix `.nodl.xml`.
This XML file is exported to the [ament index][ament_index], either manually in the case of python projects or with a helper CMake macro.
The interface may cover only a subset of nodes in a package, as long as the nodes that _are_ covered are done so completely.

Here is an example NoDL for a package containing two nodes:

{% include_relative ros2_nodl/interface_declaration.xml %}

Once an NoDL file is written, it is exported from either `CMakeLists.txt` or `setup.py` (more details below).
Note that several NoDL files can be exported, allowing for writing one NoDL file per node if desired.

### Exporting a NoDL to the Ament Index

Per the design philosophy of the [ament index][ament_index], files will be installed in two locations.
In the case of a package named `Foo`, the NoDL file `foo.nodl.xml` should be placed in the package share directory, `share/foo/foo.nodl.xml`.
A corresponding marker file, `share/ament_index/nodl_desc/foo`, should be created.

> This is either empty or contains the relative path to the `foo.nodl.xml` file.

Reviewer 2: ```suggestion
It is either empty or contains the relative path to the `foo.nodl.xml` file.
```

Author: 663747e

#### CMake Macro

> For packages using ament_cmake, the package `ament_nodl` provides the macro `nodl_export_node_description_file` which performs the above.

Reviewer 2: Replace `the above` with a link to section `Exporting a NoDL to the Ament Index`. Something along the line,
```suggestion
For packages using ament_cmake, the package `ament_nodl` provides the macro `nodl_export_node_description_file` which performs the export described in [Exporting a NoDL to the Ament Index][#exporting-a-nodl-to-the-ament-index]
```

Author: 663747e

> For a package Foo, containing `foo.nodl.xml`, the following lines are added to `CMakeLists.txt`:

Reviewer 2: ```suggestion
For a package `Foo`, containing `foo.nodl.xml`, the following lines are added to `CMakeLists.txt`:
```

Author: 663747e

```cmake
find_package(ament_nodl REQUIRED)
nodl_export_node_description_file(foo.nodl.xml)
```

#### setup.py

> In the case of setup.py, the above must be done manually alongside the placement of the package's marker in the ament index.

Reviewer 2: Replace `the above`.

Author: 663747e

> You can re-use the same empty marker file placed in the package index.

Reviewer 2: ```suggestion
One can re-use the same empty marker file placed in the package index.
```

Author: 663747e
